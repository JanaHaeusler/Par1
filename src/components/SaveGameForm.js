import useFormData from '../hooks/useFormData'
import styled from 'styled-components/macro'
import Button from './Button'

export default function SaveGameForm() {
   
    const { gameProfile, setGameProfile, savedGameProfiles, setSavedGameProfiles } = useFormData()

    return (
        <FormWrapper onSubmit={handleSubmit}>
            <InputWrapper>
                <label htmlFor="location">
                    Location
                </label>
                <input 
                    type="text" 
                    name="location"
                    id="location"
                    placeholder="Type location ..."
                    value={gameProfile.location}
                    onChange={handleChange}
                />
                <label htmlFor="date">
                    Date
                </label>
                <input 
                    type="date" 
                    name="date"
                    id="date"
                    value={gameProfile.date}
                    onChange={handleChange}
                />
                <label htmlFor="players">
                    Player(s)
                </label>
                <input 
                    type="text" 
                    name="players"
                    id="players"
                    placeholder="John, Jane"
                    value={gameProfile.players}
                    onChange={handleChange}
                />
                <label htmlFor="winner">
                    Winner(s)
                </label>
                <input 
                    type="text" 
                    name="winner"
                    id="winner"
                    placeholder="Jane"
                    value={gameProfile.winner}
                    onChange={handleChange}
                />
                <label htmlFor="shots">
                    Total Shots Winner(s)
                </label>
                <input 
                    type="text"
                    name="shots"
                    id="shots"
                    placeholder="38"
                    value={gameProfile.shots}
                    onChange={handleChange}
                />
            </InputWrapper>
            <Button>&#10003; Save</Button>
        </FormWrapper>
    )


    function handleChange(event) {
        setGameProfile({
            ...gameProfile,
            [event.target.name]: event.target.value
        })
    }

    function handleSubmit(event) {
        event.preventDefault()
        setSavedGameProfiles([
            gameProfile,
            ...savedGameProfiles
        ])
        const form = event.target
        setGameProfile({
            location: '',
            date: '',
            players: '',
            winner: '',
            shots:'',
        })
        form.elements.location.focus()
    }

}

const FormWrapper = styled.form`
    padding: 0 10px 10px 10px;
    display: grid;
    place-items: center;
    border-radius: 10px;
    box-shadow: 0 0 10px lightgrey;
    font-size: 1rem;

    Button {
        margin-top: 15px;
        margin-bottom: 5px;
    }
`
const InputWrapper = styled.fieldset`
    margin: 0;
    padding: 0;
    border: none;
    display: grid;
    width: 100%;

    label {
        margin-top: 10px;
    }

    input {
        border-style: none;
        border: 1px solid var(--primary);
        border-radius: 5px;
        padding: 5px;
       
    }

`
