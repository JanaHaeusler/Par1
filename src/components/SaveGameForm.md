```js
<SaveGameForm
  gameProfile={{
    location: '',
    date: '',
    players: '',
    winner: '',
    shots: '',
  }}
  setGameProfile={() => {}}
  savedGameProfiles={[
    {
      location: '',
      date: '',
      players: '',
      winner: '',
      shots: '',
    },
    {
      location: '',
      date: '',
      players: '',
      winner: '',
      shots: '',
    },
  ]}
  setSavedGameProfiles={() => {}}
/>
```
