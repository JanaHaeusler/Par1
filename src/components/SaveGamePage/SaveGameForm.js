import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import useForm from './hooks/useForm'
import {Check, Cross} from '../Icons'
import ButtonPrimary from '../buttons/ButtonPrimary'
import ButtonSecondary from '../buttons/ButtonSecondary'

SaveGameForm.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    isEditFormShown: PropTypes.bool,
    targetProfile: PropTypes.object,
    editGameProfile: PropTypes.func,
    cancelEditModus: PropTypes.func,
    showGameCardsPage: PropTypes.func.isRequired,
}

export default function SaveGameForm({
    onSubmit, 
    isEditFormShown, 
    targetProfile, 
    editGameProfile, 
    cancelEditModus,
    showGameCardsPage}) {

    const {
        formInputs, 
        showSaveButton,
        updateDirtyInputs,
        handleChange,
        showErrorMessage,
        handleSubmit,
        handleCancelEditModus} = useForm({
            onSubmit,
            isEditFormShown,
            targetProfile,
            editGameProfile,
            cancelEditModus,
            showGameCardsPage})

    return (
        <FormWrapper noValidate onSubmit={handleSubmit}>
            <InputWrapper>
                <label>
                    Location
                <input 
                    type="text" 
                    name="location"
                    id="location"
                    placeholder="Type location ..."
                    value={formInputs.location}
                    onChange={handleChange}
                    onBlur={() => updateDirtyInputs('location')}
                />
                </label>
                <span>{showErrorMessage('location')}</span>
                
                <label>
                    Date
                <input 
                    type="date" 
                    name="date"
                    id="date"
                    value={formInputs.date}
                    onChange={handleChange}
                    onBlur={() => updateDirtyInputs('date')}
                />
                </label>
                <span>{showErrorMessage('date')}</span>

                <label>
                    Player(s)
                <input 
                    type="text" 
                    name="players"
                    id="players"
                    placeholder="John, Jane"
                    value={formInputs.players}
                    onChange={handleChange}
                    onBlur={() => updateDirtyInputs('players')}
                />
                </label>
                <span>{showErrorMessage('players')}</span>
                
                <label>
                    Winner(s)
                <input 
                    type="text" 
                    name="winner"
                    id="winner"
                    placeholder="Jane"
                    value={formInputs.winner}
                    onChange={handleChange}
                    onBlur={() => updateDirtyInputs('winner')}
                />
                </label>
                <span>{showErrorMessage('winner')}</span>

                <label>
                    Total Shots Winner(s)
                <input 
                    type="number"
                    name="shots"
                    id="shots"
                    placeholder="38"
                    value={formInputs.shots}
                    onChange={handleChange}
                    onBlur={() => updateDirtyInputs('shots')}
                />
                </label>
                <span>{showErrorMessage('shots')}</span>

            </InputWrapper>
            <ButtonWrapper>
                <ButtonPrimary disabled={!showSaveButton} data-testid="button-save"><CheckIcon/>Save</ButtonPrimary>
                {isEditFormShown && <ButtonCancel onClick={handleCancelEditModus} data-testid="button-cancel"><CancelIcon/>Cancel</ButtonCancel>}
            </ButtonWrapper>
            <span>*Please do not clear your browsers cache, in order to permanently save your game details</span>
        </FormWrapper>
    )
}

const FormWrapper = styled.form`
    padding: 0 10px 10px 10px;
    display: grid;
    place-items: center;
    border-radius: 10px;
    box-shadow: 0 0 10px var(--primary-medium);
    background-color: var(--text-light);
    color: var(--text-dark);
    font-size: 1rem;

    Button {
        margin-top: 15px;
        margin-bottom: 5px;
    }

    span {
        margin-top: 5px;
        font-size: 0.7rem;
        color: var(--text-dark);
    }
`
const InputWrapper = styled.fieldset`
    margin: 0;
    padding: 0;
    border: none;
    display: grid;
    width: 100%;

    label {
        margin-top: 10px;
        margin-bottom: 5px;
        font-family: 'Raleway', sans-serif;
    }

    input {
        display: block;
        margin-top: 3px;
        padding: 5px;
        width: 100%;
        border-style: none;
        border-bottom: 1px solid var(--primary-dark);
        color: var(--primary-dark);
        font-family: 'Montserrat', sans-serif;
    }

    span {
        margin-top: 5px;
        font-size: 0.7rem;
        color: var(--text-dark);
    }
`
const ButtonWrapper = styled.div`
    display: flex;
`
const CheckIcon = styled(Check)`
    margin-right: 3px;
    width: 24%;
    fill: var(--text-light);
`
const ButtonCancel = styled(ButtonSecondary)`
    margin-left: 20px;
`
const CancelIcon = styled(Cross)`
    margin-right: 3px;
    fill: var(--secondary-dark);
`