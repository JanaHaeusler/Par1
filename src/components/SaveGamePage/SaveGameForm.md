Normal Form

```jsx
<SaveGameForm onSubmit={() => {}} showGameCardsPage={() => {}} />
```

Form in Edit Modus

```jsx
<SaveGameForm
  onSubmit={() => {}}
  isEditFormShown={true}
  targetProfile={{
    location: 'Horner Racecourse',
    date: '2020-12-24',
    players: 'John, Jane',
    winner: 'Jane',
    shots: '89',
    _id: '1',
  }}
  editGameProfile={() => {}}
  cancelEditModus={() => {}}
  showGameCardsPage={() => {}}
/>
```
