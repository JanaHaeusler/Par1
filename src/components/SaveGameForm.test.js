import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import SaveGameForm from './SaveGameForm'

describe('SaveGameForm', () => {
  const onSubmitMock = jest.fn()

  it('calls onSubmit with correct data and resets', () => {
    const { getByLabelText, getByRole } = render(
      <SaveGameForm onSubmit={onSubmitMock} />
    )

    user.type(getByLabelText('Location'), 'foo')
    user.type(getByLabelText('Date'), '2020-11-21')
    user.type(getByLabelText('Player(s)'), 'John, Jane')
    user.type(getByLabelText('Winner(s)'), 'Jane')
    user.type(getByLabelText('Total Shots Winner(s)'), '42')

    user.click(getByRole('button'))

    expect(onSubmitMock).toHaveBeenCalledWith({
      location: 'foo',
      date: '2020-11-21',
      players: 'John, Jane',
      winner: 'Jane',
      shots: '42',
    })

    expect(getByLabelText('Location')).toHaveValue('')
    expect(getByLabelText('Date')).toHaveValue('')
    expect(getByLabelText('Player(s)')).toHaveValue('')
    expect(getByLabelText('Winner(s)')).toHaveValue('')
    expect(getByLabelText('Total Shots Winner(s)')).toHaveValue('')
  })
})
