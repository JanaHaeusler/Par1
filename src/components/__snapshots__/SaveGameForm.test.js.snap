// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveGameForm renders DOM correctly 1`] = `
.c2 {
  padding: 7px;
  border-style: none;
  border-radius: 5px;
  box-shadow: 3px 3px 10px lightgrey;
  background-color: var(--primary);
  font-size: 1rem;
}

.c0 {
  padding: 0 10px 10px 10px;
  display: grid;
  place-items: center;
  border-radius: 10px;
  box-shadow: 0 0 10px lightgrey;
  font-size: 1rem;
}

.c0 Button {
  margin-top: 15px;
  margin-bottom: 5px;
}

.c1 {
  margin: 0;
  padding: 0;
  border: none;
  display: grid;
  width: 100%;
}

.c1 label {
  margin-top: 10px;
}

.c1 input {
  border-style: none;
  border: 1px solid var(--primary);
  border-radius: 5px;
  padding: 5px;
}

<form
  class="c0"
>
  <fieldset
    class="c1"
  >
    <label
      for="location"
    >
      Location
    </label>
    <input
      id="location"
      name="location"
      placeholder="Type location ..."
      type="text"
      value=""
    />
    <label
      for="date"
    >
      Date
    </label>
    <input
      id="date"
      name="date"
      type="date"
      value=""
    />
    <label
      for="players"
    >
      Player(s)
    </label>
    <input
      id="players"
      name="players"
      placeholder="John, Jane"
      type="text"
      value=""
    />
    <label
      for="winner"
    >
      Winner(s)
    </label>
    <input
      id="winner"
      name="winner"
      placeholder="Jane"
      type="text"
      value=""
    />
    <label
      for="shots"
    >
      Total Shots Winner(s)
    </label>
    <input
      id="shots"
      name="shots"
      placeholder="38"
      type="text"
      value=""
    />
  </fieldset>
  <button
    class="c2"
  >
    âœ“ Save
  </button>
</form>
`;
